### ISSUE TITLE
implement haproxy back end checks for kubernetes rancher ingress lb

### ISSUE THREAD

**Issue by aemneina** [rancher/rancher#10102]
I've run into the situation where a docker host goes bad. The bad host houses a pod that sits behind an ingress. A new pod gets created to meet scale demands on the k8s side but the old pod ip remains in the haproxy config. Ideally we can stop sending traffic to the defunct pod.

**Steps to Reproduce:**
1. have an ingress fronting a service (with scale 2 in my case)
2. have the host housing one of the pods go bad (render host defunct)
3. 3rd pod comes up, haproxy has 3 IP's for backend service
4. 2/3'rds of the requests to backend succeed

**Comment by prachidamle** [comment]
Adding a new label 'io.rancher.kubernetes.livenessprobes' in the ingress annotations field to configure haproxy healthchecks for the backends by creating a custom config.

The format of the 'io.rancher.kubernetes.livenessprobes' value should provide details of healthcheck to be added, using k8s http or tcp livenessprobes fields:

- k8s http liveness probe: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-http-request

- k8s tcp liveness probe: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-tcp-liveness-probe

Example:

- To add http healthcheck to haproxy backend:

io.rancher.kubernetes.livenessprobes: "{"livenessProbes": [{"probe":{"service": "<service_name>", "httpGet": {"path": "", "port": <service_port>},"periodSeconds": 3, "timeoutSeconds": 4, "successThreshold": 2, "failureThreshold": 3}}]}"

- To add tcp healthcheck to haproxy backend:

io.rancher.kubernetes.livenessprobes: "{"livenessProbes": [{"probe":{"service": "<service_name>", "tcpSocket": {"port": <port_for_tcp_check>},"periodSeconds": 3, "timeoutSeconds": 3, "successThreshold": 1, "failureThreshold": 2}}]}"

- Multiple probes can also be defined for multiple services of the ingress spec

- Example k8s service/replication controller specs and ingress specs:


```
Service & RC spec: 

apiVersion: v1
kind: Service
metadata:
  name: test-http
  labels:
    k8s-app: test-http
spec:
  ports:
  - port: 91
    targetPort: 80
    protocol: TCP
    name: http
  selector:
    k8s-app: test-http
---
apiVersion: v1
kind: ReplicationController
metadata:
  name: test-http
spec:
  replicas: 2
  selector:
    k8s-app: test-http
  template:
    metadata:
      labels:
        k8s-app: test-http
    spec:
      containers:
      - name: test-http
        image: nginx
        ports:
          - containerPort: 80
---
ingress spec with http healthcheck: 

apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: test-http-simplelb
  annotations:
    livenessprobes: "{\"livenessProbes\": [{\"probe\":{\"service\": \"test-http\", \"httpGet\": {\"path\": \"/index.html\", \"port\": 80},\"periodSeconds\": 3, \"timeoutSeconds\": 1, \"successThreshold\": 2, \"failureThreshold\": 3}}]}"   
spec:
  backend:
    serviceName: test-http
    servicePort: 91

---
ingress spec with tcp healthcheck: 

apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: test-tcp-simplelb
  annotations:
    livenessprobes: "{\"livenessProbes\": [{\"probe\":{\"service\": \"test-http\", \"tcpSocket\": {\"port\": 80},\"periodSeconds\": 3, \"timeoutSeconds\": 1, \"successThreshold\": 3, \"failureThreshold\": 4}}]}"   
spec:
  backend:
    serviceName: test-http
    servicePort: 91

```

**Comment by deniseschannon** [comment]
Available with `rancher/lb-service-rancher:v0.7.16`

**Comment by soumyalj** [comment]
Tested with v1.6.11-rc10 with rancher/lb-service-rancher:v0.7.16
On a 3 host K8s setup, created a pods and ingress with liveness probe.
Stop docker/Power OFF the host where a pod exists and a the pod was recreated on a different host.The ingress did not have multiple IPs listed in the backend service
When the ingress was accessed using wget, response was received from the targets with out any error or delay.
Also, tested with master using rancher/lb-service-rancher:v0.7.17
Ingress using liveness probe with single and multiple services as backends was also functional.

**Comment by robikovacs** [comment]
@deniseschannon how can I specify the image ```rancher/lb-service-rancher:v0.7.16 ``` for the LB using Kubernetes Ingress? I can't find any documentation

**Comment by deniseschannon** [comment]
@robikovacs  It's part of the k8s template. what version of k8s are you using? this is where it's set in the latest k8s template available with Rancher v1.6.14.

https://github.com/rancher/rancher-catalog/blob/v1.6-release/infra-templates/k8s/41/docker-compose.yml.tpl#L8

**Comment by robikovacs** [comment]
I'm using `rancher/k8s:v1.8.3-rancher3`. Thanks for clarifying @deniseschannon, so I should update Rancher to the latest version.
